Unable to process reports after a schema change

**Describe the bug**
ReportStream breaks when reports have been ingested and queued for processing, but a deployment happens that changes a schema used for the queued report.  This is due to the addition and/or deletion of fields in the schema and the internal CSV format not matching what was expected when the report was first ingested.  Note there is no impact to any reports ingested AFTER the deployment.

**Impact**
Reports are no longer processed and manual intervention is needed to resubmit the reports that failed

**To Reproduce**
Steps to reproduce the behavior locally:
1. Submit a report and stop the processing of the report, so the queue message is not pulled.
2. Stop ReportStream.
3. Change the covid-19 schema to add a field somewhere in the middle of the schema.
4. Restart ReportStream using the new schema.
5. Verify that exceptions are shown in the log that the report cannot be processed.

**Expected behavior**
Changes to the schemas after a report has been received, but not processed, should not affect the processing of the data.

**Screenshots**
N/A

**Logs**
```
Exception while executing function: Functions.process Result: Failure
Exception: IndexOutOfBoundsException: Index 81 out of bounds for length 81
Stack: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.microsoft.azure.functions.worker.broker.JavaMethodInvokeInfo.invoke(JavaMethodInvokeInfo.java:22)
	at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.execute(EnhancedJavaMethodExecutorImpl.java:55)
	at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.invokeMethod(JavaFunctionBroker.java:62)
	at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:33)
	at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:10)
	at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:45)
	at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:92)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IndexOutOfBoundsException: Index 81 out of bounds for length 81
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at gov.cdc.prime.router.Report.createTable$valuesToColumns(Report.kt:315)
	at gov.cdc.prime.router.Report.createTable(Report.kt:319)
	at gov.cdc.prime.router.Report.<init>(Report.kt:246)
	at gov.cdc.prime.router.Report.<init>(Report.kt:229)
	at gov.cdc.prime.router.serializers.CsvSerializer.readInternal(CsvSerializer.kt:181)
	at gov.cdc.prime.router.serializers.CsvSerializer.readInternal$default(CsvSerializer.kt:157)
	at gov.cdc.prime.router.azure.WorkflowEngine$handleProcessEvent$1.invoke(WorkflowEngine.kt:545)
	at gov.cdc.prime.router.azure.WorkflowEngine$handleProcessEvent$1.invoke(WorkflowEngine.kt:539)
	at gov.cdc.prime.router.azure.DatabaseAccess.transact$lambda-0(DatabaseAccess.kt:77)
	at org.jooq.impl.DefaultDSLContext.lambda$transaction$5(DefaultDSLContext.java:611)
	at org.jooq.impl.DefaultDSLContext.lambda$transactionResult0$3(DefaultDSLContext.java:549)
	at org.jooq.impl.Tools$4$1.block(Tools.java:5252)
	at java.base/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3128)
	at org.jooq.impl.Tools$4.get(Tools.java:5249)
	at org.jooq.impl.DefaultDSLContext.transactionResult0(DefaultDSLContext.java:597)
	at org.jooq.impl.DefaultDSLContext.transactionResult(DefaultDSLContext.java:521)
	at org.jooq.impl.DefaultDSLContext.transaction(DefaultDSLContext.java:610)
	at gov.cdc.prime.router.azure.DatabaseAccess.transact(DatabaseAccess.kt:77)
	at gov.cdc.prime.router.azure.WorkflowEngine.handleProcessEvent(WorkflowEngine.kt:539)
	at gov.cdc.prime.router.azure.ProcessFunction.run(ProcessFunction.kt:41)
	... 16 more
```
**Desktop (please complete the following information):**
N/A

**Smartphone (please complete the following information):**
N/A

**Additional context**

--- a/prime-router/src/main/kotlin/azure/CovidResultMetaDataFunction.kt
+++ b/prime-router/src/main/kotlin/azure/CovidResultMetaDataFunction.kt
@@ -142,8 +142,7 @@ class CovidResultMetaDataFunction : Logging {
             schema.name,
             ByteArrayInputStream(bytes),
             emptyList(),
-            blobReportId = reportFile.reportId,
-            useDefaultsForMissing = true
+            blobReportId = reportFile.reportId
         )
     }
 

--- a/prime-router/src/main/kotlin/azure/WorkflowEngine.kt
+++ b/prime-router/src/main/kotlin/azure/WorkflowEngine.kt
@@ -659,8 +659,7 @@ class WorkflowEngine(
                     ByteArrayInputStream(bytes),
                     emptyList(),
                     header.receiver,
-                    header.reportFile.reportId,
-                    true
+                    header.reportFile.reportId
                 )
             }
             else -> error("Unsupported read format")

--- a/prime-router/src/main/kotlin/cli/ProcessDataCommands.kt
+++ b/prime-router/src/main/kotlin/cli/ProcessDataCommands.kt
@@ -266,8 +266,7 @@ class ProcessData(
                     csvSerializer.readInternal(
                         schema.name,
                         file.inputStream(),
-                        listOf(FileSource(file.nameWithoutExtension)),
-                        useDefaultsForMissing = true
+                        listOf(FileSource(file.nameWithoutExtension))
                     )
                 } else {
                     val result =

--- a/prime-router/src/main/kotlin/serializers/CsvSerializer.kt
+++ b/prime-router/src/main/kotlin/serializers/CsvSerializer.kt
@@ -160,14 +160,11 @@ class CsvSerializer(val metadata: Metadata) : Logging {
         destination: Receiver? = null,
         blobReportId: ReportId? = null
     ): Report {
-        // find our schema
         val schema = metadata.findSchema(schemaName) ?: error("Internal Error: invalid schema name '$schemaName'")
-        // get our rows
         val rows = csvReader().readAllWithHeader(input).map {
             // For each element name, if it doesn't exist in the map, then we add it with a default.
-            // Doing it this way means that even if someone adds a new element in the middle of the schema
-            // after a report has been received, but before completely process, then this should still be okay
-            // and it won't necessarily break
+            // This is so we can add/remove fields/elements anywhere in a schema and not break the process step
+            // when a report has been queued and its schema changes before processing.
             schema.elements.map { element ->
                 it.getOrDefault(element.name, element.default ?: "")
             }

--- a/prime-router/src/main/kotlin/serializers/CsvSerializer.kt
+++ b/prime-router/src/main/kotlin/serializers/CsvSerializer.kt
@@ -147,36 +147,30 @@ class CsvSerializer(val metadata: Metadata) : Logging {
     }
 
     /**
-     * Reads an internal format document (one that is in CSV, but matching the internal "kitchen sink" schema)
-     * and returns the Report object for the file.
-     * @param useDefaultsForMissing if a field is missing that the Report object expects, this instructs the
-     * function to use the default value provided by the schema (or empty string) if it doesn't exist. This was
-     * added because some older files might need to be reprocessed, and this allows that to happen without error
-     * due to a missing column
+     * Reads an internal format document from [input] using the schema [schemaName]and returns the Report object for
+     * the report for a given [destination]
+     * @param sources the list of sources for this report
+     * @param blobReportId the report ID
+     * @return the report
      */
     fun readInternal(
         schemaName: String,
         input: InputStream,
         sources: List<Source>,
         destination: Receiver? = null,
-        blobReportId: ReportId? = null,
-        useDefaultsForMissing: Boolean = false
+        blobReportId: ReportId? = null
     ): Report {
         // find our schema
         val schema = metadata.findSchema(schemaName) ?: error("Internal Error: invalid schema name '$schemaName'")
         // get our rows
-        val rows = if (useDefaultsForMissing) {
-            // walk through the rows in the file with the header included
-            csvReader().readAllWithHeader(input).map {
-                // for each element name, if it doesn't exist in the map, then we add it with a default
-                // doing it this ways means that even if someone adds a new element in the middle of the schema
-                // (please don't), this should still be okay and it won't necessarily break
-                schema.elements.map { element ->
-                    it.getOrDefault(element.name, element.default ?: "")
-                }
+        val rows = csvReader().readAllWithHeader(input).map {
+            // For each element name, if it doesn't exist in the map, then we add it with a default.
+            // Doing it this way means that even if someone adds a new element in the middle of the schema
+            // after a report has been received, but before completely process, then this should still be okay
+            // and it won't necessarily break
+            schema.elements.map { element ->
+                it.getOrDefault(element.name, element.default ?: "")
             }
-        } else {
-            csvReader().readAll(input).drop(1)
         }
         return Report(schema, rows, sources, destination, id = blobReportId, metadata = metadata)
     }

--- a/prime-router/src/test/kotlin/serializers/CsvSerializerTests.kt
+++ b/prime-router/src/test/kotlin/serializers/CsvSerializerTests.kt
@@ -594,4 +594,47 @@ class CsvSerializerTests {
         }
         assertThat(err.details).isNotEmpty()
     }
+
+    @Test
+    fun `test schema changes do not affect reading`() {
+        val schema = Schema(
+            name = "one",
+            topic = "test",
+            elements = listOf(
+                Element("a", csvFields = Element.csvFields("a")),
+                Element("c", csvFields = Element.csvFields("c"))
+            )
+        )
+        val serializer = CsvSerializer(Metadata(schema))
+
+        // Internal CSV with extra fields.
+        var internalCsv = """
+            a,b,c
+            a1,b1,c1
+            a2,b2,c2
+        """.trimIndent()
+        var report = serializer.readInternal(
+            schema.name, ByteArrayInputStream(internalCsv.toByteArray()),
+            emptyList()
+        )
+        assertThat(report.itemCount).isEqualTo(2)
+        assertThat(report.getString(0, "a")).isEqualTo("a1")
+        assertThat(report.getString(0, "c")).isEqualTo("c1")
+        assertThat(report.getString(1, "c")).isEqualTo("c2")
+
+        // Internal CSV with fewer fields.
+        internalCsv = """
+            c
+            c1
+            c2
+        """.trimIndent()
+        report = serializer.readInternal(
+            schema.name, ByteArrayInputStream(internalCsv.toByteArray()),
+            emptyList()
+        )
+        assertThat(report.itemCount).isEqualTo(2)
+        assertThat(report.getString(0, "a")).isEqualTo("")
+        assertThat(report.getString(0, "c")).isEqualTo("c1")
+        assertThat(report.getString(1, "c")).isEqualTo("c2")
+    }
 }

--- a/prime-router/src/testIntegration/kotlin/datatests/TranslationTests.kt
+++ b/prime-router/src/testIntegration/kotlin/datatests/TranslationTests.kt
@@ -257,8 +257,7 @@ class TranslationTests {
                         CsvSerializer(metadata).readInternal(
                             schema.name,
                             input,
-                            listOf(TestSource),
-                            useDefaultsForMissing = true
+                            listOf(TestSource)
                         )
                     }
                     else -> {

